presto-main/src/main/java/com/facebook/presto/cost/CachingStatsProvider.java

```diff
@@ -14,17 +14,28 @@
 package com.facebook.presto.cost;
 
 import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
```

Comment:
Winning `~12ms` on planning 6 joins, and winning `130ms` on planning 8 joins seems to be negligible. A query that joins 6 (8) tables has no chance to be sub-second.

What i'm also worried about, is that the `CachingStatsProvider` has no eviction mechanism. But maybe that's not a problem at the moment. Not sure how often do we have plans that are morphing a lot of times during the interative planning.

Reply:
@kokosing can you comment in improvements it brings?

@arhimondr AFAIR, `CachingStatsProvider` lives during rule application, so eviction is not needed. Other than that, stats are cached in Memo (for the duration of iterative optimizer invocation), but this is nothing changed within this PR. And those stats are discarded whenever "equivalence group" member changes.

Reply:
It is about 20%, so it is is hard to say if it is much or not. Also this depends on how complex join expressions are. Notice that in future it may change, for example I would expect that using histograms could require more time to calculate stats and costs.

Reply:
@kokosing i added a comment in `EquivalentJoinCache` explaining we are not exploring the full join spectrum:
```
// Note: plans sharing the same CacheKey describe the same relation. However, calculated stats may differ,
// e.g. due to NDV capping when reducing JOIN is applied before a JOIN that multiplies rows.
// Here we ignore those differences for performance reasons.
```

Didn't we have cases when the improvement was substantially more than 20%?

Reply:
@findepi 

> AFAIR, CachingStatsProvider lives during rule application, so eviction is not needed

Yes, i misread the code. You are right.

> Didn't we have cases when the improvement was substantially more than 20%?

It seems that even 100% performance boost in join reordering given the cost of scheduling of 6+ joins (7 table scans at least, 7 different split sources) is still negligible. Given the problems explained here (https://github.com/prestodb/presto/pull/11267#discussion_r211345420) i would think twice to make sure this optimization is needed. 

Anyhow, i don't feel strong about it. You can keep it if you think it makes sense.